### DISCUSSION TITLE
no nearest_points attribute for python bindings

### DISCUSSION THREAD

**Discussion by almazgaliev** [coal-library/coal#623] [Q&A] [answered]
Hi, i couldn't find `nearest_points` attribute in python using `dir` function on `CollisionResult`,

It would be useful for testing things in python, Is it possible to implement or there are some issues with it? (i have no experience for writing python bindings for cpp libraries).

**Comment by lmontaut** [comment, marked as answer]
Hi @almazgaliev,
A `CollisionResult` stores a vector of `Contact` (in case there is indeed a collision).
I don't know which branch of hpp-fcl you are using, but on the devel one (which will soon be released, @jorisv is working on it), you can retrieve the nearest points. On the master branch, you can use the normal and the contact point to find the associated "nearest" points.
On the `devel` branch, you can go over each `Contact` and use `getNearestPoint1/2` methods:
```python
col_res = hppfcl.CollisionResult()
[...] # do some collision checks
if col_res.isCollision():
    contact = col_res.getContact(0) # get the first `Contact` object
    cp1, cp2 = contact.getNearestPoint1(), contact.getNearestPoint2()
```

If you are using the latest release which corresponds to the `master` branch, you can instead do:
```python
col_res = hppfcl.CollisionResult()
[...] # do some collision checks
if col_res.isCollision():
    contact = col_res.getContact(0) # get the first `Contact` object
    cp = contact.pos
    normal = contact.normal
    d = contact.penetration_depth
    cp1, cp2 = cp - d * normal, cp + d * normal
```

A few notes:
- Using the latest devel branch fixes bugs related to inconsistent nearest points that the master branch has
- For primitive shapes (spheres, ellipsoids, capsules, cones, cylinders, convex meshes etc), each `CollisionResult` contains at most one `Contact`.

**Comment by almazgaliev** [comment]
thanks!
